pkg/kubelet/pod_workers.go

```diff
@@ -221,76 +602,396 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
+	}
+	// no matter what, we always supply a grace period of 1
+	if gracePeriod < 1 {
+		gracePeriod = 1
+	}
+	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
+}
+
+func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
+	var lastSyncTime time.Time
+	for update := range podUpdates {
+		pod := update.Options.Pod
+
+		klog.V(4).InfoS("Processing pod event", "pod", klog.KObj(pod), "podUID", pod.UID, "updateType", update.WorkType)
+		err := func() error {
+			// The worker is responsible for ensuring the sync method sees the appropriate
+			// status updates on resyncs (the result of the last sync), transitions to
+			// terminating (no wait), or on terminated (whatever the most recent state is).
+			// Only syncing and terminating can generate pod status changes, while terminated
+			// pods ensure the most recent status makes it to the api server.
+			var status *kubecontainer.PodStatus
+			var err error
+			switch {
+			case update.Options.RunningPod != nil:
+				// when we receive a running pod, we don't need status at all
+			default:
+				// wait until we see the next refresh from the PLEG via the cache (max 2s)
+				// TODO: this adds ~1s of latency on all transitions from sync to terminating
+				//  to terminated, and on all termination retries (including evictions). We should
+				//  improve latency by making the the pleg continuous and by allowing pod status
+				//  changes to be refreshed when key events happen (killPod, sync->terminating).
+				//  Improving this latency also reduces the possibility that a terminated
+				//  container's status is garbage collected before we have a chance to update the
+				//  API server (thus losing the exit code).
+				status, err = p.podCache.GetNewerThan(pod.UID, lastSyncTime)
+			}
+			if err != nil {
+				// This is the legacy event thrown by manage pod loop all other events are now dispatched
+				// from syncPodFn
+				p.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedSync, "error determining status: %v", err)
+				return err
+			}
+
+			ctx := p.contextForWorker(pod.UID)
+
+			// Take the appropriate action (illegal phases are prevented by UpdatePod)
+			switch {
+			case update.WorkType == TerminatedPodWork:
+				err = p.syncTerminatedPodFn(ctx, pod, status)
+
+			case update.WorkType == TerminatingPodWork:
+				p.acknowledgeTerminating(pod)
+				var gracePeriod *int64
+				var podStatusFn func(*v1.PodStatus)
+				if opt := update.Options.KillPodOptions; opt != nil {
+					gracePeriod = opt.PodTerminationGracePeriodSecondsOverride
+					podStatusFn = opt.PodStatusFunc
+				}
+				err = p.syncTerminatingPodFn(ctx, pod, status, update.Options.RunningPod, gracePeriod, podStatusFn)
+
+			default:
+				err = p.syncPodFn(ctx, update.Options.UpdateType, pod, update.Options.MirrorPod, status)
+			}
+
+			lastSyncTime = time.Now()
+			return err
+		}()
+
+		switch {
+		case err == context.Canceled:
```

Comment:
I understand what we are trying to do. However, the sync function doesn't use context today, having this log may be confusing if the grace period is really shortened.

Can we remove the context related change and graceful period shorten logic from this PR for now? It doesn't seem to take any effect right now, and can be added in a following up change.

Reply:
I really don't want to remove the grace period logic shortening in the prep step, since if we get an error in syncTerminatingPod and have to retry I think we should be using the shortened period.  Note that cancellation changes the core structure, so if we're going to soak I want to soak with the context in place.  I agree it isn't used, so I guess we could simplify some parts, but then we'd bring those back without the infra I expect.

I agree that having it in without testing doesn't accomplish much - I'd prefer to spend the time now adding the unit tests then for verifying context cancel requeues in the right scenarios even if no one is cancelling, vs waiting to add them since it's so fundamental to the idea that we *can* sync.  Once we start cancelling, it'll be pretty obvious where the bugs are (especially with e2es on pods with long preStop hooks), so the overall cancellation stuff benefits from having a fairly well tested core loop with one touch point.